notes/plutus-core/Serialisation/Flat.md

```diff
@@ -0,0 +1,129 @@
+
+Because the `Program` data type has only one constructor we know that flat will not waste any space serialising it. `ann` will always be (for serialised ASTs) `()`, which similarly to the `Program` data type has only one constructor and flat will not serialise it.
+
+Next, the `Version` is a tuple of 3 `Natural` numbers, which are encoded as variable length unsigned integers. Because all the version numbers can fit in a 7 bit word, we only need one byte to store each of them. Also, the first bit, which represents the list constructor will always be `0` (standing for `Nil`), resulting in:
+
+> 0 (*Last*) 000 (*Unused*) 1011   (*11 in binary*)
+> 0 (*Last*) 00  (*Unused*) 10110  (*22 in binary*)
+> 0 (*Last*) 0   (*Unused*) 100001 (*33 in binary*)
+
+## The integer constant `1`.
+
+Let's take a quick look at how we defined untyped plutus terms, in the `Language.UntypedPlutusCore.Core.Type` to see what we need to serialise.
+
+> data Term name uni fun ann
+>     = Constant ann (Some (ValueOf uni))
+>     | Builtin ann fun
+>     | Var ann name
+>     | LamAbs ann name (Term name uni fun ann)
+>     | Apply ann (Term name uni fun ann) (Term name uni fun ann)
+>     | Delay ann (Term name uni fun ann)
+>     | Force ann (Term name uni fun ann)
+>     | Error ann
+>     deriving stock (Show, Functor, Generic)
+>     deriving anyclass (NFData)
+
+We need to encode the `Constant`, signed integer value `11`. Terms are encoded using 4 bits, and the `Constant` term has tag 4 (as defined in the specification, and the `Language.UntypedPlutusCore.Core.Instance.Flat` module). This results in:
+
+> 0 (*Unused*) 100 (*4 in binary*)
+
+We find the definition of serialisation for constant kinds in the `Language.PlutusCore.Flat` haskell module:
+
+> -- See Note [The G, the Tag and the Auto].
+> instance (Closed uni, uni `Everywhere` Flat) => Flat (Some (ValueOf uni)) where
+>     encode (Some (ValueOf uni x)) =
+>       encode (Some $ TypeIn uni) <> bring (Proxy @Flat) uni (encode x)
+
+The first part is an encoding of the type of constant, and the second part is the constant itself.
+
+For the `Default` universe we have the following definitions for constant types:
+
+> -- See Note [Stable encoding of tags].
+> encodeUni DefaultUniInteger    = [0]
+> encodeUni DefaultUniByteString = [1]
+> encodeUni DefaultUniString     = [2]
+> encodeUni DefaultUniChar       = [3]
+> encodeUni DefaultUniUnit       = [4]
+> encodeUni DefaultUniBool       = [5]
+
+
+So we see how, for the integer type we care about the type is encoded as a list containing the id `0`. We know that we are using 3 bits to store the type of constant, so the encoding will be:
+
+> 1 (*Cons*) 0000 (The `0` tag using 4 bits for storage) 0 (*Nil*)
+
+The annotation will not be serialised, and we are left with the constant itself. Because it is an variable length signed integer, we first need to find out it's value after conversion to the `ZigZag` format.
```

Comment:
```suggestion
The annotation will not be serialised, and we are left with the constant itself. Because it is an variable length signed integer, we first need to find out its value after conversion to the `ZigZag` format.
```